### DISCUSSION TITLE
Compensate for bluetooth latency

### DISCUSSION THREAD

**Discussion by TobiasBales** [CoderLine/alphaTab#1454] [Ideas]
Hey,
I would (personally) love if alphaTab would/could compensate for bluetooth latency - in the browser this can be discovered via [AudioContext.outputLatency](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/outputLatency).
Would this something you would be interested in/willing to support?
If so, and if you could give me a pointer of where a good spot for that might be I'd gladly try my hand at adding support for it

**Comment by Danielku15** [comment]
That definitely makes sense. I assume with "compensation" you mean effectively that the cursor and advertised "play position" are adjusted with this given latency? 

If the `outputLatency` is not changing it should be fairly easy to add. If it is changing it might be hard to continuously adjust the position but it should be possible.

Techncially I think it makes sense to extend this event: 

* https://github.com/CoderLine/alphaTab/blob/develop/src/synth/ISynthOutput.ts#L58

Instead of just passing on the number of samples played, it could pass on a structure containing samples and latency. 
```ts
class SamplesPlayedEventArgs {
    public samplesPlayed:number;
    public outputLatency:number;
}
```

Then in the main synth the latency has to be handled: 

* https://github.com/CoderLine/alphaTab/blob/develop/src/synth/AlphaSynth.ts#L405-L413

I'd have to also fiddle a bit around to see where exactly the latency should be respected to not break anything on the synthesizer itself. We might need to add new properties to differenciate the "synthesizing" time line and the "heard" timeline. 

* https://github.com/CoderLine/alphaTab/blob/34ad6af50eca0c36cccab2113eeac3ef47985070/src/synth/AlphaSynth.ts#L459-L501

I added https://github.com/CoderLine/alphaTab/issues/1455 as work item to track this feature and added some thoughts there. .
